feat(core,common,testing): support overriding middleware for testing

schiemon opened the pull request:
closes: #4073

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features) I will update them if this PR is ready to merge.


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

## What is the current behavior?
The `TestingModuleBuilder` in the `testing` package currently has no method available to override middleware à la `overrideProvider` or `overrideModule`.

Issue Number: #4073 

## What is the new behavior?

Now, it is possible to override middleware as it is configured by the `configure` method of a module:

```typescript
@Module({})
class AppModule {
  configure(consumer: MiddlewareConsumer) {
    return consumer
      .apply(MiddlewareA, MiddlewareB, MiddlewareC)
      .forRoutes('*');
  }
}

const testingModule = await Test.createTestingModule({
      imports: [AppModule],
})
  .overrideMiddleware(MiddlewareA)
  .useMiddleware(MiddlewareAOverride)
  .overrideMiddleware(MiddlewareC)
  .useMiddleware(MiddlewareC1Override, MiddlewareC2Override)
  .compile();
``` 

As a by-product, the `MiddlewareBuilder` is extended by a `replace` method that allows for swapping out middleware:

```typescript
      builder
        .apply(MiddlewareA, MiddlewareB, MiddlewareC)
        .exclude(route)
        .forRoutes("*")
        .replace(MiddlewareA, MiddlewareAOverride)
        .replace(MiddlewareC, MiddlewareC1Override, MiddlewareC2Override);
```

## Does this PR introduce a breaking change?
- [ ] Yes
- [x] No

## Other information

kamilmysliwiec commented on packages/testing/interfaces/override-middleware.interface.ts:
```diff
@@ -0,0 +1,12 @@
+ * @publicApi
+ */
+export interface OverrideMiddleware {
+  useMiddleware: (
```
```suggestion
  use: (
```

schiemon replied:
I did this because of the `overrideModule`/`useModule`-Pair but I guess looking at this from the consumer side it is less wordy.

schiemon commented:
I renamed the `use`-Function and just added the docs for this feature. Are there any other open points that you want me to have a look at, @kamilmysliwiec?

Tony133 commented on packages/testing/testing-module.builder.ts:
```diff
@@ -150,6 +175,21 @@ export class TestingModuleBuilder {
+    for (const { instance } of this.container.getModules().values()) {
+      const originalConfigurationMethod = instance.configure;
+      instance.configure = (middlewareConsumer: MiddlewareConsumer) => {
+        if (!originalConfigurationMethod) return [];
```
```suggestion
        if (!originalConfigurationMethod) {
          return [];
        }
```

Tony133 commented on integration/testing/middleware-override/e2e/middleware-override.spec.ts:
```diff
@@ -0,0 +1,191 @@
+        false;
+  };
+
+  beforeEach(() => {
+    resetMiddlewareApplicationFlags();
+  });
```
```suggestion
  beforeEach(() => {
    resetMiddlewareApplicationFlags();
  });
  
```

Tony133 reviewed (changes requested):
Let's leave one line space, for the rest I would say it can be fine

schiemon commented:
@kamilmysliwiec Any updates on this?

kamilmysliwiec commented on packages/common/interfaces/middleware/middleware-consumer.interface.ts:
```diff
@@ -16,4 +16,17 @@ export interface MiddlewareConsumer {
+   *
+   * @returns {MiddlewareConsumer}
+   */
+  replace(
```
I'm generally OK with this PR except for this change. Is there any chance we could get rid of this method?

schiemon replied:
To remove that method, as far as I see, we need a way to swap the `MiddlewareConsumer` instance with an implementation that can support replacements. One could accomplish that through the following changes:

First, one enables the injection of `MiddlewareModule` into a `NestApplication`:

```ts
constructor(
    container: NestContainer,
    private readonly httpAdapter: HttpServer,
    private readonly config: ApplicationConfig,
    private readonly graphInspector: GraphInspector,
    private readonly middlewareModule: MiddlewareModule = new MiddlewareModule(), // `middlewareModule` is now injected and not internally initialized
    appOptions: NestApplicationOptions = {},
) {
...
```

In the `MiddlewareModule` itself, we then allow the insertion of a custom `MiddlewareBuilder` in `loadConfiguration` like that:

```ts
  public async loadConfiguration(
    middlewareContainer: MiddlewareContainer,
    moduleRef: Module,
    moduleKey: string,
    middlewareBuilder?: MiddlewareBuilder, // We now can specify the builder we want to use to configure the modules
  ) {
   ...

    middlewareBuilder =
      middlewareBuilder ??
      new MiddlewareBuilder(
        this.routesMapper,
        this.httpAdapter,
        this.routeInfoPathExtractor,
      );
  ...
```

With that in place, we can then inject a `TestMiddlewareModule` that inherits from `MiddlewareModule` into our `NestApplication`. Our custom `TestMiddlewareModule` can then override `loadConfiguration` which we call the `loadConfiguration` from the parent `MiddlewareModule` with a custom `TestMiddlewareBuilder`. The `TestMiddlewareBuilder` inherits from the `MiddlewareBuilder`, which finally supports the `replace` method we can use.

I am unsure of the consequences of the changes on central classes like `NestApplication`, but in contrast to the `replace` method in `MiddlewareBuilder`, they make sense in the respective classes (I assume this is the reason why you want the method to be removed).

What do you (or others!) think about that?
